### ISSUE TITLE
Two questions about val.py's result.

### ISSUE THREAD

**Issue by xf-cpp** [ultralytics/yolov5#10263]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello!
I have trained coco dataset with a changed yolov5s.Then I used val.py to test the result,and I found that the result from terminal output and pycocotools is inconsistent.
Here is the result.
val: data=data\coco.yaml, weights=D:\Experiment\yolov5-6.0\runs\weights\best.pt, batch_size=32, imgsz=640, conf_thres=0.001, iou_thres=0.6, task=val, device=, single_cls=False, augment=False, verbose=False, save_txt=False, save_hybrid=False, save_conf=False, save_json=True, project=runs\val, name=coco+LDW(10+K13+14+K7)+SIou+8-Attention-9, exist_ok=False, half=False
YOLOv5  2022-10-27 torch 1.12.1+cu113 CUDA:0 (NVIDIA GeForce RTX 3090 Ti, 24563.5MB)
Fusing layers... 
E:\Anaconda\envs\v5s\lib\site-packages\torch\functional.py:478: UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at  C:\actions-runner\_work\pytorch\pytorch\builder\windows\pytorch\aten\src\ATen\native\TensorShape.cpp:2895.)
  return _VF.meshgrid(tensors, **kwargs)  # type: ignore[attr-defined]
Model Summary: 217 layers, 7614237 parameters, 0 gradients, 16.8 GFLOPs
val: Scanning 'D:\dateset\COCO2017\val2017.cache' images and labels... 4952 found, 48 missing, 0 empty, 0 corrupted: 100%|██████████| 5000/5000 [00:00<?, ?it/s]
Class       Images     Labels          P          R       mAP@.5   mAP@.5:.95: 100%|██████████| 157/157 [00:30<00:00,  5.14it/s]   
     all       5000      36335      0.689      0.526      0.579      0.375
Speed: 0.2ms pre-process, 1.4ms inference, 1.0ms NMS per image at shape (32, 3, 640, 640)
Evaluating pycocotools mAP... saving runs\val\best_predictions.json...
loading annotations into memory...
Done (t=0.55s)
creating index...
index created!
Loading and preparing results...
DONE (t=3.67s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *bbox*
DONE (t=12.76s).
Accumulating evaluation results...
DONE (t=6.46s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.001
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.002
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.001
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.001
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.002
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.003
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.037
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.066
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.071
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.053
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.072

 And when I adjust the conf_thres to 0.6,the result would become this.

val: data=data\coco.yaml, weights=D:\Experiment\yolov5-6.0\runs\train\weights\best.pt, batch_size=32, imgsz=640, conf_thres=0.6, iou_thres=0.6, task=val, device=, single_cls=False, augment=False, verbose=False, save_txt=False, save_hybrid=False, save_conf=False, save_json=True, project=runs\val, name= exist_ok=False, half=False
YOLOv5  2022-10-27 torch 1.12.1+cu113 CUDA:0 (NVIDIA GeForce RTX 3090 Ti, 24563.5MB)
Fusing layers... 
E:\Anaconda\envs\v5s\lib\site-packages\torch\functional.py:478: UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at  C:\actions-runner\_work\pytorch\pytorch\builder\windows\pytorch\aten\src\ATen\native\TensorShape.cpp:2895.)
  return _VF.meshgrid(tensors, **kwargs)  # type: ignore[attr-defined]
Model Summary: 217 layers, 7614237 parameters, 0 gradients, 16.8 GFLOPs
val: Scanning 'D:\dateset\COCO2017\val2017.cache' images and labels... 4952 found, 48 missing, 0 empty, 0 corrupted: 100%|██████████| 5000/5000 [00:00<?, ?it/s]
 Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100%|██████████| 157/157 [00:22<00:00,  6.90it/s]
                 all       5000      36335      0.842      0.365      0.609      0.469
Speed: 0.2ms pre-process, 1.4ms inference, 0.6ms NMS per image at shape (32, 3, 640, 640)
Evaluating pycocotools mAP... saving runs\val\best_predictions.json...
loading annotations into memory...
Done (t=0.40s)
creating index...
index created!
Loading and preparing results...
DONE (t=0.04s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *bbox*
DONE (t=3.64s).
Accumulating evaluation results...
DONE (t=0.88s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.001
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.001
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.001
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.001
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000

I don't know what causesd this.If you know why,please tell me.
Thanks a lot!

### Additional

Here is some enviroments infomation.
pycocotools 2.0.6
torch            1.12.1+cu113
torchaudio   0.12.1+cu113
torchextractor 0.3.0
torchvision 0.13.1+cu113

**Comment by Laughing-q** [comment]
@xiafneg It's possible that your category index is not correct to COCO. The category range of coco is 0~91, but in yolov5 training  it's 0\~80. So we have to map the category index back to 0\~91 when doing coco evaluation. Make sure the `is_coco` equals to `True` when doing coco evaluation then the `class_map` works.
[https://github.com/ultralytics/yolov5/blob/7398d2d77cbac9f66259926d49c26bfa3c257a9b/val.py#L189](https://github.com/ultralytics/yolov5/blob/7398d2d77cbac9f66259926d49c26bfa3c257a9b/val.py#L189)

**Comment by xiafneg** [comment]
@Laughing-q Thanks a lot! You're my hero!
I checked the `is_coco`,and  I found it is False.The reason is that my filepath is `COCO2017\2017val.txt`,and the default filepath is `coco\val2017.txt'`.
Here is the result.

val: data=data\coco.yaml, weights=D:\Experiment\yolov5-6.0\runs\train\weights\best.pt, batch_size=32, imgsz=640, conf_thres=0.001, iou_thres=0.6, task=val, device=, single_cls=False, augment=False, verbose=False, save_txt=False, save_hybrid=False, save_conf=False, save_json=True, project=runs\val, name=test, exist_ok=False, half=False
YOLOv5  2022-10-27 torch 1.12.1+cu113 CUDA:0 (NVIDIA GeForce RTX 3090 Ti, 24563.5MB)
Fusing layers... 
E:\Anaconda\envs\v5s\lib\site-packages\torch\functional.py:478: UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at  C:\actions-runner\_work\pytorch\pytorch\builder\windows\pytorch\aten\src\ATen\native\TensorShape.cpp:2895.)
  return _VF.meshgrid(tensors, **kwargs)  # type: ignore[attr-defined]
Model Summary: 217 layers, 7614237 parameters, 0 gradients, 16.8 GFLOPs
val: Scanning 'D:\dateset\COCO2017\val2017.cache' images and labels... 4952 found, 48 missing, 0 empty, 0 corrupted: 100%|██████████| 5000/5000 [00:00<?, ?it/s]
Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100%|██████████| 157/157 [00:30<00:00,  5.23it/s]
all            5000      36335      0.689      0.526      0.579      0.375
Speed: 0.2ms pre-process, 1.4ms inference, 1.0ms NMS per image at shape (32, 3, 640, 640)
Evaluating pycocotools mAP... saving runs\val\test9\best_predictions.json...
loading annotations into memory...
Done (t=0.55s)
creating index...
index created!
Loading and preparing results...
DONE (t=3.65s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *bbox*
DONE (t=30.34s).
Accumulating evaluation results...
DONE (t=7.38s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.384
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.585
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.413
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.213
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.430
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.508
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.314
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.520
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.571
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.376
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.633
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.725

But the results from terminal output and the pycocotool are still inconsistent.Do you know how to fix this?And does the map calculated with val.py during training affect the training process？

**Comment by xiafneg** [comment]
@Laughing-q Thanks again!
I adjust the `iou-thres` to 0.65,and there is still a gap between the results which I mentioned above.

val: data=data\coco.yaml, weights=D:\Experiment\yolov5-6.0\runs\train\weights\best.pt, batch_size=32, imgsz=640, conf_thres=0.001, iou_thres=0.65, task=val, device=, single_cls=False, augment=False, verbose=False, save_txt=False, save_hybrid=False, save_conf=False, save_json=True, project=runs\val, name=test, exist_ok=False, half=False
YOLOv5  2022-10-27 torch 1.12.1+cu113 CUDA:0 (NVIDIA GeForce RTX 3090 Ti, 24563.5MB)
Fusing layers... 
E:\Anaconda\envs\v5s\lib\site-packages\torch\functional.py:478: UserWarning: torch.meshgrid: in an upcoming release, it will be required to pass the indexing argument. (Triggered internally at  C:\actions-runner\_work\pytorch\pytorch\builder\windows\pytorch\aten\src\ATen\native\TensorShape.cpp:2895.)
  return _VF.meshgrid(tensors, **kwargs)  # type: ignore[attr-defined]
Model Summary: 217 layers, 7614237 parameters, 0 gradients, 16.8 GFLOPs
val: Scanning 'D:\dateset\COCO2017\val2017.cache' images and labels... 4952 found, 48 missing, 0 empty, 0 corrupted: 100%|██████████| 5000/5000 [00:00<?, ?it/s]
Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95: 100%|██████████| 157/157 [00:30<00:00,  5.12it/s]
  all       5000      36335      0.698      0.519      0.576      0.373
Speed: 0.2ms pre-process, 1.4ms inference, 1.0ms NMS per image at shape (32, 3, 640, 640)
Evaluating pycocotools mAP... saving runs\val\test11\best_predictions.json...
loading annotations into memory...
Done (t=0.35s)
creating index...
index created!
Loading and preparing results...
DONE (t=4.36s)
creating index...
index created!
Running per image evaluation...
Evaluate annotation type *bbox*
DONE (t=30.47s).
Accumulating evaluation results...
DONE (t=7.49s).
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.385
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.582
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.417
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.214
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.431
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.509
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.314
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.523
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.579
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.381
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.640
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.733

Anyway,thanks for helping my problem.Have a nice day.
